cli/debug: collect namespaces, descriptors, and jobs with AOST

fqazi opened the pull request:
Previously, when collecting these tables for debug doctor or the debug.zip we did not use AOST queries. This could be problematic when there was contention against these tables, leading to retries. To address this, these tables will now always be collected with AOST.

Fixes: #103608

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103807)
<!-- Reviewable:end -->

postamar commented on pkg/cli/zip_table_registry.go:
```diff
@@ -1037,12 +1037,18 @@ var zipSystemTables = DebugZipTableRegistry{
+"name",
+"id"
+FROM system.namespace AS OF SYSTEM TIME '-0.1s'
+`,
```
I think the idea is that all tables which get dumped should be queried AOST and not just the catalog tables, but I may be wrong.

fqazi replied:
We can opt instead for ```BEGIN AS OF SYSTEM TIME ...``` and have this blanket applied within the debug zip command

postamar replied:
Works for me

fqazi commented on pkg/cli/doctor.go:
```diff
@@ -212,16 +212,16 @@ func fromCluster(
 SELECT id, descriptor, crdb_internal_mvcc_timestamp AS mod_time_logical
-FROM system.descriptor ORDER BY id`
-	checkColumnExistsStmt := "SELECT crdb_internal_mvcc_timestamp FROM system.descriptor LIMIT 1"
+FROM system.descriptor  AS OF SYSTEM TIME '-0.1s' ORDER BY id`
```
Good point, and we should use a constant here. The only painful thing here is that from a unit testing viewpoint, AOST queries with longer intervals will mean that we need to modify our tests to sleep/wait for that long. Otherwise, we might have a time travel query before the existence of any tables.

postamar replied:
Ah, right. Can we have this be a global var which gets overridden by an `envutil.EnvOrDefault...` using an env var which is only set in test settings, or something? Presumably there's machinery which exists to do this kind of thing, though I'm not familiar with it. Perhaps @tbg or @knz can enlighten us?

A fraction of a second makes sense for tests. Perhaps this is also fine in prod, idk, in any case it's better than what we presently have, I suppose.

postamar replied:
Can either of these have default values which are different depending on whether we're running tests or not?

chengxiong-ruan replied:
oh I think NO, but sorry that didn't make the point clear. I was wondering if it's possible that users would want to change the setting. But the odds is pretty low. And for testing, I we could just set the setting while a db handle is created, but if we could achieve from env var then I agree that's better.

postamar replied:
I definitely agree with you that they might want to, and perhaps even disable the AOST thing altogether.

fqazi replied:
I went with the buildutil flag to distinguish the cases.

fqazi committed:
cli/debug: collect namespaces, descriptors, and jobs with AOST

Previously, when collecting these tables for debug doctor or
the debug.zip we did not use AOST queries. This could be
problematic when there was contention against these tables,
leading to retries. To address this, these tables will now
always be collected with AOST.

Fixes: #103608

Release note: None

fqazi commented:
@postamar TFTR!

fqazi commented:
bors r+
